zerver/tests/test_message_send.py

```diff
@@ -2412,27 +2415,144 @@ def test_personal(self) -> None:
 
         bot_profile = self.create_test_bot("testbot", user_profile)
         notification_bot = get_system_bot("notification-bot@zulip.com", user_profile.realm_id)
+        # Tests if messages to and from bots are allowed irrespective of the value of the setting.
         self.send_personal_message(user_profile, notification_bot)
         self.send_personal_message(user_profile, bot_profile)
         self.send_personal_message(bot_profile, user_profile)
 
+        # Tests if the permission works when the setting is set to a combination of
+        # groups and users.
+        user_group = self.create_or_update_anonymous_group_for_setting(
+            [user_profile],
+            [administrators_system_group],
+        )
+        do_change_realm_permission_group_setting(
+            realm,
+            "direct_message_initiator_group",
+            user_group,
+            acting_user=None,
+        )
+        self.send_personal_message(user_profile, cordelia)
+
+    def test_direct_message_permission_group_setting(self) -> None:
+        """
+        Tests that direct_message_permission_group_setting works correctly.
+        """
+        user_profile = self.example_user("hamlet")
+        cordelia = self.example_user("cordelia")
+        admin = self.example_user("iago")
+        realm = user_profile.realm
+        huddle = [user_profile, cordelia, admin]
+        huddle_without_admin = [user_profile, cordelia]
+        administrators_system_group = NamedUserGroup.objects.get(
+            name=SystemGroups.ADMINISTRATORS, realm=realm, is_system_group=True
+        )
+        self.login_user(user_profile)
+        do_change_realm_permission_group_setting(
+            realm,
+            "direct_message_permission_group",
+            administrators_system_group,
+            acting_user=None,
+        )
+        # Tests if the user is allowed to send to administrators.
+        with self.assert_database_query_count(16):
+            self.send_personal_message(user_profile, admin)
+        self.send_personal_message(admin, user_profile)
+        # Tests if we can send messages to self irrespective of the value of the setting.
+        self.send_personal_message(user_profile, user_profile)
+
+        # We cannot send direct messages unless one of the recipient is in the
+        # `direct_message_permission_group` (in this case, the
+        # `administrators_system_group`).
+        with self.assertRaises(JsonableError):
+            self.send_personal_message(user_profile, cordelia)
+
+        # We can send to this direct message group as it has administrator as one of the
+        # recipient.
+        with self.assert_database_query_count(24):
+            self.send_huddle_message(user_profile, huddle)
+        self.send_huddle_message(admin, huddle)
+
+        # But this one does not have an administrator. So, it should throw an error.
+        with self.assertRaises(JsonableError):
```

Comment:
Probably these asserts should include actually verifying what error occurred; otherwise, if the test setup was broken and these started failing for some other reason (like `huddle_without_admin` being invalid in some way), we'd not know.

One theory for how to do that cleanly is you could make a named subclass of `JsonableError` for this class of permission failure, defined in `zerver/lib/exceptions.py`. I don't think we necessarily need a `CodedError`.